File: server/src/queue/download.rs
use crate::db_optimized::Db;
use crate::storage::{get_today_folder, THUMBNAILS_ROOT};
use regex::Regex;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::{Arc, Mutex};
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio_util::sync::CancellationToken;

use super::DownloadQueue;

impl DownloadQueue {
    pub async fn generate_thumbnail(video_path: &Path) -> Result<PathBuf, anyhow::Error> {
        let filename = video_path
            .file_name()
            .ok_or_else(|| anyhow::anyhow!("No filename"))?;
        let thumb_filename = format!("{}.jpg", filename.to_string_lossy());
        let thumb_path = Path::new(THUMBNAILS_ROOT).join(thumb_filename);

        if !Path::new(THUMBNAILS_ROOT).exists() {
            let _ = tokio::fs::create_dir_all(THUMBNAILS_ROOT).await;
        }
        if thumb_path.exists() {
            return Ok(thumb_path);
        }

        let mut cmd = Command::new("ffmpeg");
        cmd.arg("-i")
            .arg(video_path)
            .arg("-ss")
            .arg("00:00:01")
            .arg("-vframes")
            .arg("1")
            .arg("-q:v")
            .arg("4")
            .arg(&thumb_path)
            .stdout(Stdio::null())
            .stderr(Stdio::null());

        let status = cmd.status().await?;
        if !status.success() {
            let mut cmd = Command::new("ffmpeg");
            cmd.arg("-i")
                .arg(video_path)
                .arg("-ss")
                .arg("00:00:00")
                .arg("-vframes")
                .arg("1")
                .arg("-q:v")
                .arg("4")
                .arg(&thumb_path)
                .stdout(Stdio::null())
                .stderr(Stdio::null());
            let status = cmd.status().await?;
            if !status.success() {
                return Err(anyhow::anyhow!("ffmpeg failed to generate thumbnail"));
            }
        }

        Ok(thumb_path)
    }

    pub(super) async fn run_yt_dlp(
        id: &str,
        url: &str,
        category: &str,
        db: &Db,
        token: CancellationToken,
    ) -> Result<(String, Option<String>, Option<String>, Option<String>), anyhow::Error> {
        let job = db.get_job(id).await?;
        let user_role = if let Some(uid) = &job.user_id {
            let user: Result<crate::db_optimized::models::User, _> = db.db.collection("users")
                .find_one(mongodb::bson::doc! { "username": uid }).await
                .map(|o| o.unwrap_or_else(|| crate::db_optimized::models::User { 
                    id: "".to_string(), username: "guest".to_string(), email: "".to_string(), 
                    password_hash: "".to_string(), role: "guest".to_string(), default_preset_id: None 
                }));
            user.map(|u| u.role).unwrap_or_else(|_| "guest".to_string())
        } else {
            "guest".to_string()
        };

        let cwd = std::env::current_dir()?;
        let python_path = std::env::var("YT_DLP_PYTHON")
            .map(PathBuf::from)
            .unwrap_or_else(|_| cwd.join("venv_python/bin/python"));
        let yt_dlp_path = std::env::var("YT_DLP_BINARY")
            .map(PathBuf::from)
            .unwrap_or_else(|_| cwd.join("bin/yt-dlp"));

        let output_folder = get_today_folder(Some(category));
        let is_tiktok = url.contains("tiktok.com");
        let is_instagram = url.contains("instagram.com");
        let template = if is_instagram || is_tiktok {
            output_folder.join("%(id)s.%(ext)s")
        } else {
            output_folder.join("%(title)s.%(ext)s")
        };

        let mut cmd = Command::new("nice");
        cmd.arg("-n")
            .arg("10")
            .arg(&python_path)
            .arg(&yt_dlp_path)
            .arg("--newline")
            .arg("--no-check-certificates")
            .arg("--no-mtime")
            .arg("--no-update")
            .arg("--js-runtimes")
            .arg("node");

        // --- Quality & Preset Logic ---
        let mut quality_applied = false;

        // 1. Check for Custom Preset (Premium Only)
        if user_role == "premium_member" {
            if let Some(pid) = job.preset_id {
                if let Ok(preset) = db.get_preset(&pid).await {
                    for arg in preset.args {
                        if crate::validation::is_safe_ytdlp_arg(&arg) {
                            cmd.arg(arg);
                        } else {
                            tracing::warn!("Skipping unsafe custom preset argument: {}", arg);
                        }
                    }
                    quality_applied = true;
                }
            }
        }

        // 2. Default Role-Based Quality
        if !quality_applied {
            if user_role == "admin" {
                // Admin: Absolute highest quality available
                cmd.arg("-f").arg("bv*+ba/best");
            } else if user_role == "premium_member" {
                // Premium Default: High quality
                cmd.arg("-f").arg("bv*+ba/best");
            } else {
                // Guest: Cap at 1080p and 1GB file size
                cmd.arg("-f").arg("bv*[height<=1080]+ba/b[height<=1080] / bestvideo+bestaudio / best");
                cmd.arg("--max-filesize").arg("1G");
            }
        }

        cmd.arg("--merge-output-format")
            .arg("mp4")
            .arg("--remux-video")
            .arg("mp4")
            .arg("--postprocessor-args")
            .arg("ffmpeg:-movflags +faststart")
            .arg("--write-info-json");

        if is_tiktok {
            cmd.arg("--add-header")
                .arg("Referer:https://www.tiktok.com/");
        } else if is_instagram {
            cmd.arg("--add-header")
                .arg("Referer:https://www.instagram.com/");
        }

        // Auto-detect cookie files for platforms that require auth
        let cookie_file = if is_instagram && Path::new("cookies_instagram.txt").exists() {
            Some("cookies_instagram.txt")
        } else if url.contains("youtube.com") || url.contains("youtu.be") {
            if Path::new("cookies_youtube.txt").exists() {
                Some("cookies_youtube.txt")
            } else if Path::new("cookies.txt").exists() {
                Some("cookies.txt")
            } else {
                None
            }
        } else if Path::new("cookies.txt").exists() {
            Some("cookies.txt")
        } else {
            None
        };

        if let Some(cf) = cookie_file {
            tracing::info!("Using cookie file: {}", cf);
            cmd.arg("--cookies").arg(cf);
        }

        let actual_url = if is_tiktok && url.contains("/photo/") {
            url.replace("/photo/", "/video/")
        } else {
            url.to_string()
        };

        let mut child = cmd
            .arg("-o")
            .arg(template)
            .arg("--")
            .arg(&actual_url)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .spawn()?;

        let stdout = child.stdout.take().expect("Failed to open stdout");
        let stderr = child.stderr.take().expect("Failed to open stderr");

        let found_filename = Arc::new(Mutex::new(String::new()));
        let found_filename_clone = found_filename.clone();
        let db_clone = db.clone();
        let id_clone = id.to_string();

        let stdout_task = tokio::spawn(async move {
            let mut reader = BufReader::new(stdout).lines();
            let mut last_progress_update = std::time::Instant::now();
            let re_progress = Regex::new(r"[download]\s+(\d+\.?\d*?)%").unwrap();
            let re_eta = Regex::new(r"ETA\s+(\d{2}:\d{2}(?:\:\d{2})?)").unwrap();
            let re_dest = Regex::new(r"\b[dD]estination:\s+(.*)").unwrap();
            let re_merge = Regex::new(r#"\b[mM]erger\b.*into\s+\"?([^\"]*)\"?"#).unwrap();
            let re_already =
                Regex::new(r"\b[dD]ownloaded\s+(.*)\s+has already been downloaded").unwrap();

            while let Ok(Some(line)) = reader.next_line().await {
                if let Some(caps) = re_progress.captures(&line) {
                    if let Some(m) = caps.get(1) {
                        if let Ok(p) = m.as_str().parse::<f64>() {
                            if last_progress_update.elapsed().as_secs() >= 1 {
                                let eta = re_eta
                                    .captures(&line)
                                    .and_then(|eta_caps| eta_caps.get(1))
                                    .and_then(|m| Self::parse_eta(m.as_str()));
                                let _ = db_clone.update_progress(&id_clone, p as i64, eta).await;
                                last_progress_update = std::time::Instant::now();
                            }
                        }
                    }
                }

                if let Some(caps) = re_dest.captures(&line) {
                    if let Some(m) = caps.get(1) {
                        let mut w = found_filename_clone.lock().unwrap();
                        *w = m.as_str().trim().to_string();
                    }
                }

                if let Some(caps) = re_merge.captures(&line) {
                    if let Some(m) = caps.get(1) {
                        let mut w = found_filename_clone.lock().unwrap();
                        *w = m.as_str().trim().trim_matches('"').to_string();
                    }
                }

                if let Some(caps) = re_already.captures(&line) {
                    if let Some(m) = caps.get(1) {
                        {
                            let mut w = found_filename_clone.lock().unwrap();
                            *w = m.as_str().trim().to_string();
                        }
                        let _ = db_clone.update_progress(&id_clone, 100, Some(0)).await;
                    }
                }
            }
        });

        let stderr_error = Arc::new(Mutex::new(String::new()));
        let stderr_error_clone = stderr_error.clone();

        let stderr_task = tokio::spawn(async move {
            let mut reader = BufReader::new(stderr).lines();
            while let Ok(Some(line)) = reader.next_line().await {
                // Capture ERROR lines from yt-dlp for better error reporting
                if line.contains("ERROR:") {
                    let mut w = stderr_error_clone.lock().unwrap();
                    if w.is_empty() {
                        *w = line.trim().to_string();
                    }
                }
            }
        });

        tokio::select! {
            _ = token.cancelled() => {
                child.kill().await?;
                Err(anyhow::anyhow!("Job cancelled"))
            }
            status = child.wait() => {
                let status = status?;
                let _ = stdout_task.await;
                let _ = stderr_task.await;

                if status.success() {
                    let mut full_path_str = found_filename.lock().unwrap().clone();
                    
                    if full_path_str.is_empty() {
                        // Fallback: find the newest .mp4 file in the output folder
                        if let Ok(mut entries) = tokio::fs::read_dir(&output_folder).await {
                            let mut newest_file = None;
                            let mut newest_time = std::time::SystemTime::UNIX_EPOCH;
                            while let Ok(Some(entry)) = entries.next_entry().await {
                                if let Ok(meta) = entry.metadata().await {
                                    if let Some(ext) = entry.path().extension() {
                                        if ext == "mp4" || ext == "mkv" || ext == "webm" {
                                            if let Ok(modified) = meta.modified() {
                                                if modified > newest_time {
                                                    newest_time = modified;
                                                    newest_file = Some(entry.path());
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                            if let Some(p) = newest_file {
                                full_path_str = p.to_string_lossy().to_string();
                            }
                        }
                    }

                    if !full_path_str.is_empty() {
                        let path = Path::new(&full_path_str);
                        let filename = path.file_name().unwrap().to_string_lossy().to_string();
                        let json_path = if let Some(stem) = path.file_stem() {
                            path.with_file_name(format!("{}.info.json", stem.to_string_lossy()))
                        } else {
                            PathBuf::from(&full_path_str).with_extension("info.json")
                        };

                        let mut creator = None;
                        let avatar = None;
                        let mut caption = None;

                        if json_path.exists() {
                            if let Ok(content) = tokio::fs::read_to_string(&json_path).await {
                                if let Ok(json) = serde_json::from_str::<serde_json::Value>(&content) {
                                    creator = json.get("uploader").and_then(|v| v.as_str()).map(|s| s.to_string());
                                    caption = json.get("description").and_then(|v| v.as_str()).map(|s| s.to_string());
                                    if caption.is_none() {
                                        caption = json.get("title").and_then(|v| v.as_str()).map(|s| s.to_string());
                                    }
                                }
                            }
                            let _ = tokio::fs::remove_file(&json_path).await;
                        }

                        Ok((filename, creator, avatar, caption))
                    } else {
                        Ok(("unknown.mp4".to_string(), None, None, None))
                    }
                } else {
                    // Use captured stderr error if available for a more descriptive message
                    let stderr_msg = stderr_error.lock().unwrap().clone();
                    if !stderr_msg.is_empty() {
                        Err(anyhow::anyhow!("{}", stderr_msg))
                    } else {
                        Err(anyhow::anyhow!("Process exited with code {}", status.code().unwrap_or(-1)))
                    }
                }
            }
        }
    }

    pub(super) fn parse_eta(eta_str: &str) -> Option<i64> {
        let parts: Vec<&str> = eta_str.split(':').collect();
        let seconds = if parts.len() == 3 {
            parts[0].parse::<i64>().unwrap_or(0) * 3600
                + parts[1].parse::<i64>().unwrap_or(0) * 60
                + parts[2].parse::<i64>().unwrap_or(0)
        } else if parts.len() == 2 {
            parts[0].parse::<i64>().unwrap_or(0) * 60 + parts[1].parse::<i64>().unwrap_or(0)
        } else {
            parts[0].parse::<i64>().unwrap_or(0)
        };
        Some(seconds)
    }
}

File: server/queue/download.go
package queue

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"tiak-server/db"
	"tiak-server/models"
	"tiak-server/storage"
)

type DownloadQueue struct {
	db              *db.MongoDB
	fileIndex       *storage.FileIndex
	queue           chan string
	activeJobs      sync.Map
	maxConcurrent   int
	MaxRetryCount   uint32
	SyncDestination string
	SyncMode        string
	syncRunning     bool
	syncError       string
	syncLogs        []string
	mu              sync.RWMutex
}

func NewDownloadQueue(d *db.MongoDB, fi *storage.FileIndex, maxRetry uint32) *DownloadQueue {
	return &DownloadQueue{
		db:              d,
		fileIndex:       fi,
		queue:           make(chan string, 1000),
		maxConcurrent:   2,
		MaxRetryCount:   maxRetry,
		SyncDestination: "onedrive:EDITS",
		SyncMode:        "copy",
		syncLogs:        []string{},
	}
}

func (dq *DownloadQueue) Start(ctx context.Context) {
	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			case id := <-dq.queue:
				dq.processJob(ctx, id)
			}
		}
	}()
}

func (dq *DownloadQueue) AddJob(id string) {
	select {
	case dq.queue <- id:
	default:
	}
}

func (dq *DownloadQueue) CancelJob(id string) {
	if cancel, ok := dq.activeJobs.LoadAndDelete(id); ok {
		cancel.(context.CancelFunc)()
	}
}

func (dq *DownloadQueue) processJob(ctx context.Context, id string) {
	job, err := dq.db.GetJob(ctx, id)
	if err != nil || job.Status != "queued" {
		return
	}
	dq.db.MarkDownloading(ctx, id)

	jCtx, cancel := context.WithCancel(ctx)
	dq.activeJobs.Store(id, cancel)
	defer func() {
		dq.activeJobs.Delete(id)
		cancel()
	}()

	result, err := dq.runYtDlp(jCtx, job)
	if err != nil {
		if jCtx.Err() != nil {
			dq.db.MarkFailed(ctx, id, "Cancelled")
		} else {
			dq.db.MarkFailed(ctx, id, err.Error())
		}
		return
	}

	// Generate thumbnail
	videoPath := filepath.Join(storage.GetTodayFolder(job.Category), result.filename)
	thumbPath := filepath.Join(storage.ThumbnailsRoot, result.filename+".jpg")
	generateThumbnail(thumbPath, videoPath)

	dq.db.MarkDone(ctx, id, result.filename, result.creator, result.avatar, result.caption)
	dq.fileIndex.AddFile(videoPath)
}

type downloadResult struct {
	filename string
	creator  *string
	avatar   *string
	caption  *string
}

func (dq *DownloadQueue) runYtDlp(ctx context.Context, job *models.Job) (*downloadResult, error) {
	downloader, downloaderArgs, err := resolveYtDlpCommand()
	if err != nil {
		return nil, err
	}

	outputFolder := storage.GetTodayFolder(job.Category)
	isTikTok := strings.Contains(job.URL, "tiktok.com")
	isInstagram := strings.Contains(job.URL, "instagram.com")

	template := filepath.Join(outputFolder, "%(id)s.%(ext)s")
	if !isTikTok && !isInstagram {
		template = filepath.Join(outputFolder, "%(title)s.%(ext)s")
	}

	commandArgs := append([]string{"-n", "10", downloader}, downloaderArgs...)
	commandArgs = append(commandArgs,
		"--newline", "--no-mtime", "--no-update",
		"--merge-output-format", "mp4", "--remux-video", "mp4",
		"--postprocessor-args", "ffmpeg:-movflags +faststart", "--write-info-json",
		"-f", "bv*[height<=1080]+ba/b[height<=1080]/bestvideo+bestaudio/best")
	cmd := exec.CommandContext(ctx, "nice", commandArgs...)

	if isTikTok {
		cmd.Args = append(cmd.Args, "--add-header", "Referer:https://www.tiktok.com/")
	} else if isInstagram {
		cmd.Args = append(cmd.Args, "--add-header", "Referer:https://www.instagram.com/")
		if proxy := os.Getenv("INSTAGRAM_PROXY"); proxy != "" {
			cmd.Args = append(cmd.Args, "--proxy", proxy)
		}
	}

	cwd, _ := os.Getwd()
	for _, cf := range []string{"cookies_instagram.txt", "cookies_youtube.txt", "cookies.txt"} {
		if _, stat := os.Stat(filepath.Join(cwd, cf)); stat == nil {
			cmd.Args = append(cmd.Args, "--cookies", filepath.Join(cwd, cf))
			break
		}
	}

	actualURL := job.URL
	if isTikTok && strings.Contains(actualURL, "/photo/") {
		actualURL = strings.Replace(actualURL, "/photo/", "/video/", 1)
	}
	cmd.Args = append(cmd.Args, "-o", template, "--", actualURL)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("could not capture yt-dlp output: %w", err)
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return nil, fmt.Errorf("could not capture yt-dlp errors: %w", err)
	}
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("could not start yt-dlp: %w", err)
	}

	foundFilename := ""
	reProgress := regexp.MustCompile(`\[download\]\s+(\d+\.?\d*?)%`)
	reDest := regexp.MustCompile(`[dD]estination:\s+(.*)`)
	reMerge := regexp.MustCompile(`[mM]erger.*into\s+"?([^"]*)"?`)
	reAlready := regexp.MustCompile(`[dD]ownloaded\s+(.*)\s+has already been downloaded`)

	var outputWg sync.WaitGroup
	outputWg.Add(2)
	go func() {
		defer outputWg.Done()
		scanner := bufio.NewScanner(stdout)
		for scanner.Scan() {
			line := scanner.Text()
			if m := reProgress.FindStringSubmatch(line); len(m) > 1 {
				if p, e := strconv.ParseFloat(m[1], 64); e == nil {
					dq.db.UpdateJobProgress(ctx, job.ID, int64(p), nil)
				}
			}
			if m := reDest.FindStringSubmatch(line); len(m) > 1 {
				foundFilename = filepath.Base(strings.TrimSpace(m[1]))
			}
			if m := reMerge.FindStringSubmatch(line); len(m) > 1 {
				foundFilename = filepath.Base(strings.Trim(strings.TrimSpace(m[1]), "\""))
			}
			if m := reAlready.FindStringSubmatch(line); len(m) > 1 {
				foundFilename = filepath.Base(strings.TrimSpace(m[1]))
				dq.db.UpdateJobProgress(ctx, job.ID, 100, int64Ptr(0))
			}
		}
	}()

	stderrOutput := ""
	go func() {
		defer outputWg.Done()
		var lines []string
		scanner := bufio.NewScanner(stderr)
		for scanner.Scan() {
			lines = append(lines, strings.TrimSpace(scanner.Text()))
		}
		stderrOutput = strings.TrimSpace(strings.Join(lines, "\n"))
		const maxErrorLength = 12000
		if len([]rune(stderrOutput)) > maxErrorLength {
			runes := []rune(stderrOutput)
			stderrOutput = "…" + string(runes[len(runes)-maxErrorLength+1:])
		}
	}()

	err = cmd.Wait()
	outputWg.Wait()
	if err != nil {
		if stderrOutput != "" {
			return nil, fmt.Errorf("%s", stderrOutput)
		}
		return nil, fmt.Errorf("yt-dlp failed: %v", err)
	}

	if foundFilename == "" {
		entries, _ := os.ReadDir(outputFolder)
		newest := ""
		newestTime := time.Time{}
		for _, e := range entries {
			ext := filepath.Ext(e.Name())
			if ext == ".mp4" || ext == ".mkv" || ext == ".webm" {
				if info, eErr := e.Info(); eErr == nil && info.ModTime().After(newestTime) {
					newestTime = info.ModTime()
					newest = e.Name()
				}
			}
		}
		if newest != "" {
			foundFilename = newest
		}
	}

	if foundFilename == "" {
		foundFilename = "unknown.mp4"
	}

	// Parse info.json
	var creator, caption *string
	fullPath := filepath.Join(outputFolder, foundFilename)
	jsonPath := strings.TrimSuffix(fullPath, filepath.Ext(fullPath)) + ".info.json"
	if data, readErr := os.ReadFile(jsonPath); readErr == nil {
		content := string(data)
		if idx := strings.Index(content, `"uploader"`); idx > -1 {
			after := content[idx:]
			if colonIdx := strings.Index(after, ":"); colonIdx > -1 {
				val := strings.Trim(after[colonIdx+1:], " \"\n,")
				if endIdx := strings.IndexAny(val, "\",\n"); endIdx > -1 {
					c := val[:endIdx]
					creator = &c
				}
			}
		}
		if idx := strings.Index(content, `"description"`); idx > -1 {
			after := content[idx:]
			if colonIdx := strings.Index(after, ":"); colonIdx > -1 {
				val := strings.Trim(after[colonIdx+1:], " \"\n,")
				if endIdx := strings.IndexAny(val, "\",\n"); endIdx > -1 {
					c := val[:endIdx]
					caption = &c
				}
			}
		}
		os.Remove(jsonPath)
	}

	return &downloadResult{filename: foundFilename, creator: creator, avatar: nil, caption: caption}, nil
}

func resolveYtDlpCommand() (string, []string, error) {
	cwd, _ := os.Getwd()
	pythonPath := resolveConfiguredPath(os.Getenv("YT_DLP_PYTHON"), filepath.Join(cwd, "venv_python", "bin", "python"))
	ytDlpPath := resolveConfiguredPath(os.Getenv("YT_DLP_BINARY"), filepath.Join(cwd, "bin", "yt-dlp"))

	if isExecutable(ytDlpPath) {
		return ytDlpPath, nil, nil
	}
	if isRegularFile(ytDlpPath) && isExecutable(pythonPath) {
		return pythonPath, []string{ytDlpPath}, nil
	}
	if path, err := exec.LookPath("yt-dlp"); err == nil {
		return path, nil, nil
	}
	if isExecutable(pythonPath) {
		return pythonPath, []string{"-m", "yt_dlp"}, nil
	}

	return "", nil, fmt.Errorf("yt-dlp is not installed or configured (checked %s and %s)", pythonPath, ytDlpPath)
}

func resolveConfiguredPath(value, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		value = fallback
	}
	if filepath.IsAbs(value) {
		return value
	}
	candidates := []string{filepath.Join(mustGetwd(), value)}
	if executable, err := os.Executable(); err == nil {
		candidates = append(candidates, filepath.Join(filepath.Dir(executable), value))
	}
	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
	}
	return candidates[0]
}

func mustGetwd() string {
	cwd, err := os.Getwd()
	if err != nil {
		return "."
	}
	return cwd
}

func isRegularFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular()
}

func isExecutable(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular() && info.Mode().Perm()&0111 != 0
}

func generateThumbnail(thumbPath, videoPath string) {
	if _, err := os.Stat(thumbPath); err == nil {
		return
	}
	os.MkdirAll(storage.ThumbnailsRoot, 0755)
	cmd := exec.Command("ffmpeg", "-i", videoPath, "-ss", "00:00:01", "-vframes", "1", "-q:v", "4", thumbPath)
	cmd.Run()
}

func int64Ptr(v int64) *int64 { return &v }
